server/src/fstar_connection.ts

```diff
@@ -0,0 +1,490 @@
+	private fstar: FStar;
+	debug: boolean;
+
+	constructor(fstar: FStar, debug: boolean) {
+		this.debug = debug;
+
+		// F*'s IDE protocol requires that each request have a unique query-id.
+		// We use a monotonic id.
+		this.last_query_id = 0;
+
+		// Queries may be responded to asynchronously, so we keep a running map
+		// of pending responses to handle query responses from F*.
+		this.pending_responses = new Map();
+
+		// TODO(klinvill): Should try to spawn F* from within this constructor
+		// instead.
+		this.fstar = fstar;
+		this.fstar.proc.stdin?.setDefaultEncoding('utf-8');
+
+		// Register message handlers that will resolve the appropriate pending
+		// response promises for each query.
+		//
+		// The bufferedHandler buffers up received input until it finds a valid
+		// message. The wrapped handler will then be called with the parsed
+		// valid message. This handles receiving fragmented messages or multiple
+		// messages over the stream from the F* process.
+		const bufferedHandler = FStarConnection.bufferedMessageHandlerFactory((msg: object) => this.handleResponse(msg));
+		this.fstar.proc.stdout?.on('data', bufferedHandler);
+
+		// F* error messages are just printed out
+		const fstar_proc_name = this.fstar.lax ? 'fstar lax' : 'fstar';
+		this.fstar.proc.stderr?.on('data', (data) => { console.error(`${fstar_proc_name} stderr: ${data}`); });
+	}
+
+	// Attempts to spawn an F* process, using the given configuration and
+	// filePath, and create a connection to it.
+	//
+	// @throws {Error} from `trySpawnFstar`
+	static tryCreateFStarConnection(fstarConfig: FStarConfig, filePath: URI, debug: boolean, lax?: 'lax') : FStarConnection | undefined {
+		const fstar = FStar.trySpawnFstar(fstarConfig, filePath, debug, lax);
+		if (fstar)
+			return new FStarConnection(fstar, debug);
+		else
+			return undefined;
+	}
+
+	// Kills the F* process and closes the connection
+	close() {
+		this.fstar.proc.kill();
+	}
+
+	async restartSolver() {
+		this.fstar.killZ3SubProcess(this.debug);
+
+		// Wait for a second for processes to die before restarting the solver
+		await setTimeout(1000);
+		this.restartSolverRequest();
+	}
+
+	// Gets the configuration for the F* process.
+	fstar_config() : FStarConfig {
+		return this.fstar.config;
+	}
+
+	////////////////////////////////////////////////////////////////////////////////////
+	// Methods to send requests to F*
+	////////////////////////////////////////////////////////////////////////////////////
+
+	// Sends a request to the F* process.
+	//
+	// @param expect_response indicates if a response from F* is expected for
+	// this request.
+	private async request(msg: any, expect_response = true, is_stream = false): Promise<any> {
+		// TODO(klinvill): do we need to worry about racing requests here that might result in the same query-id? That would cause issues when tracking responses.
```

Comment:
Concurrency in javascript is a bit different than in C. :smile: 

Javascript is completely single threaded, and promises are just a nice API for cooperative multi-threading.  In other words, everything is atomic unless there's an await in the middle.